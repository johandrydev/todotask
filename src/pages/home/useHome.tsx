import { useEffect, useReducer } from "react";
import { useHistory } from "react-router";
import { useAuthContext } from "../../auth/AuthContext";
import { EActionAuth, ETodoType } from "../../types/types";

interface IAction {
  type: ETodoType;
  payload: any
};

export interface ITask {
  ID?: string;
  task: string;
  isReady: boolean;
}

const todoReducer = (state: ITask[] = [], action: IAction) => {
  switch (action.type) {
    case ETodoType.ADD:
      return [...state, action.payload];

    case ETodoType.DELETE:
      return state.filter((todo: ITask) => todo.ID !== action.payload);

    case ETodoType.TOGGLE:
      return state.map((todo: ITask) =>
        (todo.ID === action.payload)
          ? { ...todo, isReady: !todo.isReady }
          : todo
      );

    default:
      return state;
  }
};

const useHome = () => {
  const { dispatch } = useAuthContext();
  const history = useHistory();

  const [todos, changeTodo] = useReducer(todoReducer, []);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);


  const handleDelete = (todoId: string) => {
    changeTodo({
      type: ETodoType.DELETE,
      payload: todoId
    });
  };

  const handleToggle = (todoId: string) => {
    changeTodo({
      type: ETodoType.TOGGLE,
      payload: todoId
    });
  };

  const handleAddTodo = (newTodo: ITask) => {
    changeTodo({
      type: ETodoType.ADD,
      payload: newTodo
    });
  };

  const handleLogOut = () => {
    dispatch({
      type: EActionAuth.LOGOUT,
    });
    history.replace('/auth/login');
  };

  return {
    todos,
    handleDelete,
    handleToggle,
    handleAddTodo,
    handleLogOut
  };
};

export default useHome;
