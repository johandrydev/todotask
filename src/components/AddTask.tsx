import { FC, useMemo } from "react";
import { useForm } from "../hooks/useForm";
import { ITask } from "../pages/home/useHome";

const AddTask: FC<IAddTask> = ({ submit }) => {
  const [values, handleChange, resetForm] = useForm({
    task: ''
  });

  const handleSubmit = (e: any) => {
    e.preventDefault();
    if (values.task.trim().length <= 1) {
      return;
    }

    const newTodo: ITask = {
      ID: new Date().getTime().toString(),
      task: values.task,
      isReady: false
    };

    submit(newTodo);
    resetForm();
  }

  const isDisabledSubmit = useMemo(() => values?.task?.trim() === '', [values]);
  return (
    <>
      <h5>Add task</h5>
      <hr />
      <div className="form-floating mb-2">
        <textarea
          id="task"
          name="task"
          className="form-control"
          placeholder="Leave a comment here"
          onChange={handleChange}
          value={values.task}
        ></textarea>
        <label htmlFor="task">Enter the task here...</label>
      </div>
      <div className="d-flex justify-content-end">
        <button
          type="button"
          className="btn btn-outline-primary btn-block"
          onClick={handleSubmit}
          disabled={isDisabledSubmit}
        >
          Create task
        </button>
      </div>
    </>
  );
};
interface IAddTask {
  submit: (newTodo: ITask) => void;
};
export default AddTask;
