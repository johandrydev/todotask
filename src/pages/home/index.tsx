import AddTask from "../../components/AddTask";
import { TodoList } from "../../components/TodoList";
import useHome from "./useHome";

const Home = () => {
  const {
    todos,
    handleDelete,
    handleToggle,
    handleAddTodo,
    handleLogOut
  } = useHome();

  return (
    <div className="container pt-4">
      <div className="d-flex justify-content-between">
        <h3 className="display-6">To Do Tasks</h3>
        <div className="d-flex align-items-center">
          <button
            className="btn btn-danger btn-sm"
            onClick={handleLogOut}
          >
            Log Out
          </button>
        </div>
      </div>
      <hr />
      <div className="row">
        <div className="col-12 col-md-5">
          <AddTask submit={handleAddTodo} />
        </div>
        <div className="col-12 col-md-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
      </div>
    </div>
  );
};
export default Home;
