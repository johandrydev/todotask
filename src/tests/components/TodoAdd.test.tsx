import { shallow } from 'enzyme';
import AddTask from '../../components/AddTask';
import { ITask } from '../../pages/home/useHome';

describe('Test in <AddTask />', () => {
  const handleSubmit = jest.fn();
  const wrapper = shallow(
    <AddTask
      submit={handleSubmit}
    />
  );

  test('should show correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should to call to action submit', () => {
    const value = `Task's Test`;
    wrapper.find('textarea').simulate('change', {
      target: {
        value,
        name: 'task'
      }
    });
    const formSubmit: any = wrapper.find('form').prop('onSubmit');
    if (formSubmit) {
      formSubmit({ preventDefault() { } });
    }
    const todoData: ITask = {
      task: value,
      isReady: false
    }
    expect(handleSubmit).toHaveBeenCalledTimes(1);
    expect(handleSubmit).toHaveBeenCalledWith(expect.any(Object)); // { }
    expect(handleSubmit).toHaveBeenCalledWith(todoData);
    expect(wrapper.find('textarea').prop('value')).toBe('');

  });
})
