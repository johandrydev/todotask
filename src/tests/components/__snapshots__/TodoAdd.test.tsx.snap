// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <AddTask /> should show correctly 1`] = `ShallowWrapper {}`;
